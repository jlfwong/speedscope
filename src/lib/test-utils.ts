import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
