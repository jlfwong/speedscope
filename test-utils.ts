import {Profile, CallTreeNode, Frame} from './profile'

interface DumpedProfile {
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}
