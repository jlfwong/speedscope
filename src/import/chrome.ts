import {Profile, FrameInfo, CallTreeProfileBuilder} from '../lib/profile'
import {getOrInsert, lastOf} from '../lib/utils'
import {TimeFormatter} from '../lib/value-formatters'

interface TimelineEvent {
  pid: number
  tid: number
  ts: number
  ph: string
  cat: string
  name: string
  dur: number
  tdur: number
  tts: number
  args: {[key: string]: any}
}

interface PositionTickInfo {
  line: number
  ticks: number
}

interface CPUProfileCallFrame {
  columnNumber: number
  functionName: string
  lineNumber: number
  scriptId: string
  url: string
}

interface CPUProfileNode {
  callFrame: CPUProfileCallFrame
  hitCount: number
  id: number
  children?: number[]
  positionTicks?: PositionTickInfo[]
  parent?: CPUProfileNode
}

interface CPUProfile {
  startTime: number
  endTime: number
  nodes: CPUProfileNode[]
  samples: number[]
  timeDeltas: number[]
}

export function importFromChromeTimeline(events: TimelineEvent[]): Profile {
  // It seems like sometimes Chrome timeline files contain multiple CpuProfiles?
  // For now, choose the first one in the list.
  for (let event of events) {
    if (event.name == 'CpuProfile') {
      const chromeProfile = event.args.data.cpuProfile as CPUProfile
      return importFromChromeCPUProfile(chromeProfile)
    }
  }
  throw new Error('Could not find CPU profile in Timeline')
}

const callFrameToFrameInfo = new Map<CPUProfileCallFrame, FrameInfo>()
function frameInfoForCallFrame(callFrame: CPUProfileCallFrame) {
  return getOrInsert(callFrameToFrameInfo, callFrame, callFrame => {
    const name = callFrame.functionName || '(anonymous)'
    const file = callFrame.url
    const line = callFrame.lineNumber
    const col = callFrame.columnNumber
    return {
      key: `${name}:${file}:${line}:${col}`,
      name,
      file,
      line,
      col,
    }
  })
}

function shouldIgnoreFunction(functionName: string) {
  return functionName === '(root)' || functionName === '(idle)'
}

function shouldPlaceOnTopOfPreviousStack(functionName: string) {
  return functionName === '(garbage collector)' || functionName === '(program)'
}

export function importFromChromeCPUProfile(chromeProfile: CPUProfile): Profile {
  const profile = new CallTreeProfileBuilder(chromeProfile.endTime - chromeProfile.startTime)

  const nodeById = new Map<number, CPUProfileNode>()
  for (let node of chromeProfile.nodes) {
    nodeById.set(node.id, node)
  }
  for (let node of chromeProfile.nodes) {
    if (!node.children) continue
    for (let childId of node.children) {
      const child = nodeById.get(childId)
      if (!child) continue
      child.parent = node
    }
  }

  const samples: number[] = []
  const timeDeltas: number[] = []

  let elapsed = 0
  let lastNodeId = NaN

  // The chrome CPU profile format doesn't collapse identical samples. We'll do that
  // here to save a ton of work later doing mergers.
  for (let i = 0; i < chromeProfile.samples.length; i++) {
    const nodeId = chromeProfile.samples[i]
    if (nodeId != lastNodeId) {
      samples.push(nodeId)
      timeDeltas.push(elapsed)
      elapsed = 0
    }

    let timeDelta = chromeProfile.timeDeltas[i]
    if (timeDelta < 0) {
      console.warn('Substituting zero for unexpected time delta:', timeDelta, 'at index', i)
      timeDelta = 0
    }

    elapsed += timeDelta
    lastNodeId = nodeId
  }
  if (!isNaN(lastNodeId)) {
    samples.push(lastNodeId)
    timeDeltas.push(elapsed)
  }

  let prevStack: CPUProfileNode[] = []

  let value = 0
  for (let i = 0; i < samples.length; i++) {
    const timeDelta = timeDeltas[i + 1] || 0
    const nodeId = samples[i]
    let stackTop = nodeById.get(nodeId)
    if (!stackTop) continue

    // Find lowest common ancestor of the current stack and the previous one
    let lca: CPUProfileNode | null = null

    // This is O(n^2), but n should be relatively small here (stack height),
    // so hopefully this isn't much of a problem
    for (
      lca = stackTop;
      lca && prevStack.indexOf(lca) === -1;
      lca = shouldPlaceOnTopOfPreviousStack(lca.callFrame.functionName)
        ? lastOf(prevStack)
        : lca.parent || null
    ) {}

    // Close frames that are no longer open
    while (prevStack.length > 0 && lastOf(prevStack) != lca) {
      const closingNode = prevStack.pop()!
      const frame = frameInfoForCallFrame(closingNode.callFrame)
      profile.leaveFrame(frame, value)
    }

    // Open frames that are now becoming open
    const toOpen: CPUProfileNode[] = []
    for (
      let node: CPUProfileNode | null = stackTop;
      node && node != lca && !shouldIgnoreFunction(node.callFrame.functionName);
      // Place Chrome internal functions on top of the previous call stack
      node = shouldPlaceOnTopOfPreviousStack(node.callFrame.functionName)
        ? lastOf(prevStack)
        : node.parent || null
    ) {
      toOpen.push(node)
    }
    toOpen.reverse()

    for (let node of toOpen) {
      profile.enterFrame(frameInfoForCallFrame(node.callFrame), value)
    }

    prevStack = prevStack.concat(toOpen)
    value += timeDelta
  }

  // Close frames that are open at the end of the trace
  for (let i = prevStack.length - 1; i >= 0; i--) {
    profile.leaveFrame(frameInfoForCallFrame(prevStack[i].callFrame), value)
  }

  profile.setValueFormatter(new TimeFormatter('microseconds'))
  return profile.build()
}
