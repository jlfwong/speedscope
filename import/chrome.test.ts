import * as fs from 'fs'
import {dumpProfile} from '../test-utils'
import {importFromChromeCPUProfile, importFromChromeTimeline} from './chrome'

test('importFromChromeCPUProfile', () => {
  const input = fs.readFileSync('./sample/profiles/Chrome/65/simple.cpuprofile', 'utf8')
  const profile = importFromChromeCPUProfile(JSON.parse(input))
  expect(dumpProfile(profile)).toMatchSnapshot()
})

test('importFromChromeTimeline', () => {
  const input = fs.readFileSync('./sample/profiles/Chrome/65/simple-timeline.json', 'utf8')
  const profile = importFromChromeTimeline(JSON.parse(input))
  expect(dumpProfile(profile)).toMatchSnapshot()
})
