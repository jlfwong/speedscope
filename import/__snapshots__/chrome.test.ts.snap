// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromChromeCPUProfile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": -1,
      "file": "",
      "key": "(root)::-1:-1",
      "line": -1,
      "name": "(root)",
      "selfWeight": 0,
      "totalWeight": 45920,
    },
    Frame {
      "col": 0,
      "file": "",
      "key": "(anonymous)::0:0",
      "line": 0,
      "name": "(anonymous)",
      "selfWeight": 11575,
      "totalWeight": 45920,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "a::0:10",
      "line": 0,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 34345,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "b::5:10",
      "line": 5,
      "name": "b",
      "selfWeight": 0,
      "totalWeight": 16788,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "d::13:10",
      "line": 13,
      "name": "d",
      "selfWeight": 34345,
      "totalWeight": 34345,
    },
    Frame {
      "col": 10,
      "file": "",
      "key": "c::9:10",
      "line": 9,
      "name": "c",
      "selfWeight": 0,
      "totalWeight": 17557,
    },
  ],
  "stacks": Array [
    "(root);(anonymous) 11.57ms",
    "(root);(anonymous);a;b;d 16.79ms",
    "(root);(anonymous);a;c;d 17.56ms",
  ],
}
`;

exports[`importFromChromeTimeline 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": -1,
      "file": "",
      "key": "(root)::-1:-1",
      "line": -1,
      "name": "(root)",
      "selfWeight": 0,
      "totalWeight": 555107,
    },
    Frame {
      "col": -1,
      "file": "",
      "key": "(program)::-1:-1",
      "line": -1,
      "name": "(program)",
      "selfWeight": 12244,
      "totalWeight": 12244,
    },
    Frame {
      "col": -1,
      "file": "",
      "key": "(idle)::-1:-1",
      "line": -1,
      "name": "(idle)",
      "selfWeight": 509679,
      "totalWeight": 509679,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "a:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:0:10",
      "line": 0,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 33184,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "c:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:13:10",
      "line": 13,
      "name": "c",
      "selfWeight": 16197,
      "totalWeight": 16197,
    },
    Frame {
      "col": 10,
      "file": "file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "b:file:///Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:7:10",
      "line": 7,
      "name": "b",
      "selfWeight": 16987,
      "totalWeight": 16987,
    },
  ],
  "stacks": Array [
    "(root);(program) 10.16ms",
    "(root);(idle) 1.99ms",
    "(root);(program) 926.00µs",
    "(root);(idle) 85.95ms",
    "(root);(program) 133.00µs",
    "(root);(idle) 8.40ms",
    "(root);(program) 134.00µs",
    "(root);a;c 257.00µs",
    "(root);a;b 534.00µs",
    "(root);a;c 513.00µs",
    "(root);a;b 513.00µs",
    "(root);a;c 257.00µs",
    "(root);a;b 518.00µs",
    "(root);a;c 514.00µs",
    "(root);a;b 261.00µs",
    "(root);a;c 517.00µs",
    "(root);a;b 258.00µs",
    "(root);a;c 145.00µs",
    "(root);a;b 134.00µs",
    "(root);a;c 292.00µs",
    "(root);a;b 272.00µs",
    "(root);a;c 129.00µs",
    "(root);a;b 145.00µs",
    "(root);a;c 265.00µs",
    "(root);a;b 142.00µs",
    "(root);a;c 141.00µs",
    "(root);a;b 139.00µs",
    "(root);a;c 397.00µs",
    "(root);a;b 126.00µs",
    "(root);a;c 131.00µs",
    "(root);a;b 126.00µs",
    "(root);a;c 521.00µs",
    "(root);a;b 128.00µs",
    "(root);a;c 261.00µs",
    "(root);a;b 127.00µs",
    "(root);a;c 130.00µs",
    "(root);a;b 128.00µs",
    "(root);a;c 136.00µs",
    "(root);a;b 272.00µs",
    "(root);a;c 146.00µs",
    "(root);a;b 661.00µs",
    "(root);a;c 113.00µs",
    "(root);a;b 135.00µs",
    "(root);a;c 585.00µs",
    "(root);a;b 294.00µs",
    "(root);a;c 274.00µs",
    "(root);a;b 129.00µs",
    "(root);a;c 127.00µs",
    "(root);a;b 128.00µs",
    "(root);a;c 247.00µs",
    "(root);a;b 526.00µs",
    "(root);a;c 125.00µs",
    "(root);a;b 129.00µs",
    "(root);a;c 131.00µs",
    "(root);a;b 253.00µs",
    "(root);a;c 135.00µs",
    "(root);a;b 123.00µs",
    "(root);a;c 128.00µs",
    "(root);a;b 257.00µs",
    "(root);a;c 137.00µs",
    "(root);a;b 256.00µs",
    "(root);a;c 411.00µs",
    "(root);a;b 892.00µs",
    "(root);a;c 146.00µs",
    "(root);a;b 148.00µs",
    "(root);a;c 309.00µs",
    "(root);a;b 140.00µs",
    "(root);a;c 151.00µs",
    "(root);a;b 458.00µs",
    "(root);a;c 156.00µs",
    "(root);a;b 170.00µs",
    "(root);a;c 139.00µs",
    "(root);a;b 290.00µs",
    "(root);a;c 115.00µs",
    "(root);a;b 281.00µs",
    "(root);a;c 270.00µs",
    "(root);a;b 283.00µs",
    "(root);a;c 144.00µs",
    "(root);a;b 146.00µs",
    "(root);a;c 410.00µs",
    "(root);a;b 146.00µs",
    "(root);a;c 147.00µs",
    "(root);a;b 312.00µs",
    "(root);a;c 146.00µs",
    "(root);a;b 907.00µs",
    "(root);a;c 231.00µs",
    "(root);a;b 138.00µs",
    "(root);a;c 280.00µs",
    "(root);a;b 255.00µs",
    "(root);a;c 257.00µs",
    "(root);a;b 130.00µs",
    "(root);a;c 255.00µs",
    "(root);a;b 128.00µs",
    "(root);a;c 526.00µs",
    "(root);a;b 128.00µs",
    "(root);a;c 384.00µs",
    "(root);a;b 398.00µs",
    "(root);a;c 260.00µs",
    "(root);a;b 664.00µs",
    "(root);a;c 150.00µs",
    "(root);a;b 147.00µs",
    "(root);a;c 156.00µs",
    "(root);a;b 137.00µs",
    "(root);a;c 141.00µs",
    "(root);a;b 140.00µs",
    "(root);a;c 141.00µs",
    "(root);a;b 141.00µs",
    "(root);a;c 140.00µs",
    "(root);a;b 140.00µs",
    "(root);a;c 139.00µs",
    "(root);a;b 129.00µs",
    "(root);a;c 269.00µs",
    "(root);a;b 293.00µs",
    "(root);a;c 148.00µs",
    "(root);a;b 293.00µs",
    "(root);a;c 292.00µs",
    "(root);a;b 296.00µs",
    "(root);a;c 147.00µs",
    "(root);a;b 140.00µs",
    "(root);a;c 139.00µs",
    "(root);a;b 139.00µs",
    "(root);a;c 511.00µs",
    "(root);a;b 515.00µs",
    "(root);a;c 386.00µs",
    "(root);a;b 514.00µs",
    "(root);a;c 272.00µs",
    "(root);a;b 242.00µs",
    "(root);a;c 128.00µs",
    "(root);a;b 134.00µs",
    "(root);a;c 112.00µs",
    "(root);a;b 288.00µs",
    "(root);a;c 451.00µs",
    "(root);a;b 138.00µs",
    "(root);a;c 292.00µs",
    "(root);a;b 145.00µs",
    "(root);a;c 259.00µs",
    "(root);a;b 148.00µs",
    "(root);a;c 293.00µs",
    "(root);a;b 140.00µs",
    "(root);a;c 140.00µs",
    "(root);(idle) 61.61ms",
    "(root);(program) 118.00µs",
    "(root);(idle) 134.00µs",
    "(root);(program) 128.00µs",
    "(root);(idle) 16.40ms",
    "(root);(program) 125.00µs",
    "(root);(idle) 16.64ms",
    "(root);(program) 129.00µs",
    "(root);(idle) 16.72ms",
    "(root);(program) 130.00µs",
    "(root);(idle) 294.15ms",
    "(root);(program) 129.00µs",
    "(root);(idle) 7.68ms",
    "(root);(program) 131.00µs",
  ],
}
`;
